ddtrace/contrib/botocore/patch.py

```diff
@@ -52,23 +67,155 @@ def inject_trace_data_to_message_attributes(trace_data, entry):
+    Max size per event is 256kb (https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-putevent-size.html)
+    """
+    params = args[1]
+    if "Entries" not in params:
+        log.debug("Unable to inject context. The Event Bridge event had no Entries.")
+        return
+
+    for entry in params["Entries"]:
+        if "Detail" not in entry:
+            detail = {}
+            HTTPPropagator.inject(span.context, detail)
+            entry["Detail"] = json.dumps({"_datadog": detail})
+            continue
+
+        try:
+            data_size = sys.getsizeof(entry["Detail"])
+            if data_size + 512 >= 256000:
+                log.debug("Inject trace context will make Detail exceed max allowed size")
+                return
+
+            detail = json.loads(entry["Detail"])
+            dd_context = detail.get("_datadog", {})
+            HTTPPropagator.inject(span.context, dd_context)
+            detail["_datadog"] = dd_context
+            entry["Detail"] = json.dumps(detail)
+        except Exception:
+            log.warning("Unable to parse Detail and inject span context")
+
+
+def get_kinesis_data_object(data, try_b64=True):
+    # type: (str, Optional[bool]) -> Dict[str: Any]
+    """
+    :data: the data from a kinesis stream
+    :try_b64: whether we should try to decode the string as base64
+
+    The data from a kinesis stream comes as a string (could be json, base64 encoded, etc.)
+    We support injecting our trace context in the following two cases:
+        - json string
+        - base64 encoded json string
+    If it's neither of these, then we leave the message as it is.
+    Max data size per record is 1MB (https://aws.amazon.com/kinesis/data-streams/faqs/)
+    """
+
+    # check if data size will exceed max with headers
+    # the maximum data size including out headers should not exceed 1MB
+    data_size = sys.getsizeof(data)
+    if data_size + 512 >= 1000000:
+        return None
+
+    try:
+        return json.loads(data)
+    except Exception:
+        if try_b64:
+            try:
+                return get_kinesis_data_object(base64.b64decode(data.encode("ascii")).decode("ascii"), False)
+            except Exception:
+                return None
+
+        return None
+
+
+def inject_trace_to_kinesis_stream_data(record, span):
+    # type: (Dict[str, Any], Span) -> None
+    """
+    :record: contains args for the current botocore action, Kinesis record is at index 1
+    :span: the span which provides the trace context to be propagated
+
+    Inject trace headers into the Kinesis record's Data field in addition to the existing
+    data. Only possible if the existing data is JSON string or base64 encoded JSON string
+    """
+    if "Data" not in record:
+        log.debug("Unable to inject context. The kinesis stream has no data")
+        return
+
+    data = record["Data"]
+    data_obj = get_kinesis_data_object(data)
+    if data_obj:
```

Comment:
```suggestion
    if data_obj is not None:
```

An empty dict is falsey. If we were able to properly parse the data, but it just happened to be `"{}"` we could continue with the injection.